feat: Added a universal RPC attribute [MTT-7482] [MTT-7578] [MTT-7579] [MTT-7580] [MTT-7581]

ShadauxCat opened the pull request:
The new attribute is more configurable at both compile time and runtime than the existing ClientRpc and ServerRpc attributes, and also includes support for clients sending to themselves, and for optionally delaying the invocation of local RPCs until the start of the next frame to enable mutually recursive client/server RPCs to function on a host the same as on other clients (as well as mutually recursive Owner/NotOwner RPCs and other similar patterns).

This attribute supersedes ClientRpc and ServerRpc and will be the new default recommendation for creating RPCs. ClientRpc and ServerRpc will eventually be deprecated in favor of the new [Rpc] attribute.

By necessity of the feature, this also adds `NetworkManager.PeerClientIds` to allow clients to know the IDs of other connected clients, as well as `NetworkManager.OnPeerConnectedCallback` and `NetworkManager.OnPeerDisconnectCallback` events that are fired when this list is updated via messages from the server.

resolves #2718
resolves #2615
resolves #2504
resolves #2465
resolves #2448
resolves #2331
resolves #2326

## Changelog

- Added: Added a new RPC attribute, which is simply `Rpc`. 
  - This is a generic attribute that can perform the functions of both Server and Client RPCs, as well as enabling client-to-client RPCs. Includes several default targets: `Server`, `NotServer`, `Owner`, `NotOwner`, `Me`, `NotMe`, `ClientsAndHost`, and `Everyone`. Runtime overrides are available for any of these targets, as well as for sending to a specific ID or groups of IDs.
  - This attribute also includes the ability to defer RPCs that are sent to the local process to the start of the next frame instead of executing them immediately, treating them as if they had gone across the network. The default behavior is to execute immediately.
  - This attribute effectively replaces `ServerRpc` and `ClientRpc`. `ServerRpc` and `ClientRpc` remain in their existing forms for backward compatibility, but `Rpc` will be the recommended and most supported option.
- Added: `NetworkManager.OnConnectionEvent` as a unified connection event callback to notify clients and servers of all client connections and disconnections within the session
- Added: `NetworkManager.ServerIsHost` and `NetworkBehaviour.ServerIsHost` to allow a client to tell if it is connected to a host or to a dedicated server
- Changed: `NetworkManager.ConnectedClientsIds` is now accessible on the client side and will contain the list of all clients in the session, including the host client if the server is operating in host mode 

## Testing and Documentation

- Includes unit tests.
- Includes integration tests.
- Includes documentation for previously-undocumented public API entry points.

ShadauxCat committed:
feat: Added a universal RPC attribute

The new attribute is more configurable at both compile time and runtime than the existing ClientRpc and ServerRpc attributes, and also includes support for clients sending to themselves, and for optionally delaying the invocation of local RPCs until the start of the next frame to enable mutually recursive client/server RPCs to function on a host the same as on other clients (as well as mutually recursive Owner/NotOwner RPCs and other similar patterns).

This attribute supersedes ClientRpc and ServerRpc and will be the new default recommendation for creating RPCs. ClientRpc and ServerRpc will eventually be deprecated in favor of the new [Rpc] attribute.

ShadauxCat committed:
Swapping KnownClientIds to PeerClientIds

ShadauxCat committed:
Same as last commit, but better in that this one includes all the changed files.

ShadauxCat committed:
Merge branch 'develop' into feat/universal_rpc_attribute

ShadauxCat committed:
style and changelog

samlucas-unity commented on com.unity.netcode.gameobjects/Runtime/Messaging/DeferredMessageManager.cs:
```diff
@@ -113,14 +113,14 @@ public virtual void ProcessTriggers(IDeferredNetworkMessageManager.TriggerType t
                 // processed before the object is fully spawned. This must be the last thing done in the spawn process.
                 if (triggers.TryGetValue(key, out var triggerInfo))
                 {
+                    triggers.Remove(key);
```
Why the move? (for context not that I think it shouldn't move :) )

ShadauxCat replied:
If the handler for a trigger tried to add a new trigger of the same type, the `Remove()` after executing the trigger would just delete it and the trigger would never actually get executed.

samlucas-unity commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/ConnectionApprovedMessage.cs:
```diff
@@ -126,6 +148,12 @@ public void Handle(ref NetworkContext context)
             // Stop the client-side approval timeout coroutine since we are approved.
             networkManager.ConnectionManager.StopClientApprovalCoroutine();
 
+            networkManager.ConnectionManager.PeerClientIds.Clear();
```
If there any danger of these lists getting multiple of the same peer in them?  From what I've seen it doesn't look possible, but I noticed with the NetcodeObject Observers you switched it to catch duplicates, is it worth taking the precaution on the peer list?

samlucas-unity replied:
Oh its cleared 😄 right there. Well no chance of that here but I would still like your thoughts.

ShadauxCat replied:
PeerClientIds is gone now. :) But it was also a HashSet, so it would have automatically handled duplicates.

samlucas-unity commented on com.unity.netcode.gameobjects/Runtime/Messaging/RpcTargets/ClientsAndHostRpcTarget.cs:
```diff
@@ -0,0 +1,33 @@
+        {
+            if (m_UnderlyingTarget == null)
+            {
+                if (behaviour.NetworkManager.ConnectionManager.PeerClientIds.Contains(NetworkManager.ServerClientId))
```
Whats the difference here between Everyone and NotServer. Is the default to not send to self?  Why does Everyone not exclude the local client (NotMe)?  Oh is this for a host game then so in a dedicated server the ServerClientId will be null or is it just the server id.  Then this check if the server counts as a peer as in a host game so in that case the server is a client Oh its in the name ClientAndHost, so this is only to Clients and in a locally hosted game (not dedicated server) that host too.

Its clear now, maybe slap a comment on the if check to say its checking for host session?

ShadauxCat replied:
Yeah, if the server is a host, then ClientsAndHost means everyone. If the server is a dedicated server, then the server gets excluded. This is basically "send to everyone that runs any kind of client logic", which hosts do but dedicated servers don't.

samlucas-unity reviewed (approved):
None of the comments are things which need changing.  Mostly asking for context or minor suggestions.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Connection/NetworkConnectionManager.cs:
```diff
@@ -72,6 +85,7 @@ public sealed class NetworkConnectionManager
         internal Dictionary<ulong, ulong> TransportIdToClientIdMap = new Dictionary<ulong, ulong>();
         internal List<NetworkClient> ConnectedClientsList = new List<NetworkClient>();
         internal List<ulong> ConnectedClientIds = new List<ulong>();
+        internal HashSet<ulong> PeerClientIds = new HashSet<ulong>();
```
We need to discuss the peer stuff a bit more.

ShadauxCat committed:
- Merged PeerClientIds into ConnectedClientsIds
- Replaced OnPeerConnectedCallback and OnPeerDisconnectCallback with unified OnConnectionEvent

ShadauxCat committed:
Minor changelog formatting fixes, and standards fixes

ShadauxCat committed:
Updated some comments for clarity.

NoelStephensUnity commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -10,9 +10,19 @@ Additional documentation and release notes are available at [Multiplayer Documen
+  - This is a generic attribute that can perform the functions of both Server and Client RPCs, as well as enabling client-to-client RPCs. Includes several default targets: `Server`, `NotServer`, `Owner`, `NotOwner`, `Me`, `NotMe`, `ClientsAndHost`, and `Everyone`. Runtime overrides are available for any of these targets, as well as for sending to a specific ID or groups of IDs.
+  - This attribute also includes the ability to defer RPCs that are sent to the local process to the start of the next frame instead of executing them immediately, treating them as if they had gone across the network. The default behavior is to execute immediately.
+  - This attribute effectively replaces `ServerRpc` and `ClientRpc`. `ServerRpc` and `ClientRpc` remain in their existing forms for backward compatibility, but `Rpc` will be the recommended and most supported option.
+- Added `NetworkManager.OnConnectionEvent` as a unified connection event callback to notify clients and servers of all client connections and disconnections within the session (#2762)
```
:godmode: 
❤️

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -122,6 +125,11 @@ public ulong LocalClientId
+        /// <summary>
+        /// Gets whether or not the current server (local or remote) is a host - i.e., also a client
+        /// </summary>
+        public bool ServerIsHost => ConnectionManager.ConnectedClientIds.Contains(ServerClientId);
```
You could just use `ConnectionManager.LocalClient.IsHost; ` here.

ShadauxCat replied:
Only on the server side. This is used client side as well to detect if the server is connected to is a host (which is primarily relevant to figuring out how to route `SendTo.ClientsAndHost` messages, but I thought users might sometimes find this useful as well)

NoelStephensUnity replied:
Ahhhh... context is always 20:20.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -518,7 +518,10 @@ public void NetworkShow(ulong clientId)
             }
             NetworkManager.SpawnManager.MarkObjectForShowingTo(this, clientId);
-            Observers.Add(clientId);
+            if (!Observers.Contains(clientId))
```
This check isn't required since Observers is a HashSet.

ShadauxCat replied:
Oops, didn't mean to leave this in.

When trying to get clients able to send to other clients, they still have to stick to the observers list. I initially tried to do this by adding a `NetworkList<ulong> Observers` to every NetworkBehaviour and using NetworkVariable logic to handle updating it so that clients would be able to access the observer list as well, and this check was needed because NetworkList, unlike HashSet, does not handle duplicate values automatically.

That attempt broke a lot of tests, though, in ways that would be very hard to fix, so I abandoned that and just had the client send every id that could possibly match the target it is using, and had the server filter out anything that isn't in the observer list when forwarding the proxy message.

Will remove this check and look for any others I missed.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -613,7 +616,10 @@ public void NetworkHide(ulong clientId)
                     throw new VisibilityChangeException("The object is already hidden");
                 }
-                Observers.Remove(clientId);
+                if (Observers.Contains(clientId))
```
This check isn't required since Observers is a HashSet.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/ConnectionApprovedMessage.cs:
```diff
@@ -5,7 +5,8 @@ namespace Unity.Netcode
     {
-        public int Version => 0;
+        private const int k_VersionAddClientIds = 1;
+        public int Version => k_VersionAddClientIds;
```
Ahh.. we now have a new version!

ShadauxCat replied:
We do!

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/ConnectionApprovedMessage.cs:
```diff
@@ -50,7 +58,10 @@ public void Serialize(FastBufferWriter writer, int targetVersion)
                     if (sobj.SpawnWithObservers && (sobj.CheckObjectVisibility == null || sobj.CheckObjectVisibility(OwnerClientId)))
                     {
-                        sobj.Observers.Add(OwnerClientId);
+                        if (!sobj.Observers.Contains(OwnerClientId))
```
This check isn't required since Observers is a HashSet.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -69,7 +69,10 @@ internal bool RemoveObjectFromShowingTo(NetworkObject networkObject, ulong clien
             if (ret)
             {
-                networkObject.Observers.Remove(clientId);
+                if (networkObject.Observers.Contains(clientId))
```
This check isn't required since Observers is a HashSet.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -947,7 +950,10 @@ internal void UpdateObservedNetworkObjects(ulong clientId)
                     if (sobj.CheckObjectVisibility(clientId))
                     {
-                        sobj.Observers.Add(clientId);
+                        if (!sobj.Observers.Contains(clientId))
```
This check isn't required since Observers is a HashSet.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -947,7 +950,10 @@ internal void UpdateObservedNetworkObjects(ulong clientId)
+                        }
                     }
                     else // Otherwise, if the observers contains the clientId (shouldn't happen) then remove it since CheckObjectVisibility returned false
                     if (sobj.Observers.Contains(clientId))
```
We could probably remove this check too...looks like it was already there.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/Runtime/RpcTests.cs:
```diff
@@ -62,7 +62,6 @@ public void MyNativeListServerRpc(NativeList<ulong> clientId, ServerRpcParams pa
             }
 #endif
 
-
```
You could remove the file from this PR by adding the CR/LF back.  (_nit-picky of me_)

ShadauxCat replied:
When I save it in Rider, it just removes the CR/LF again. 😔

NoelStephensUnity reviewed (approved):
Looks Good To Me!
:godmode: 
Awesome! 
_(Some minor nit-picky suggestions on very minor adjustments made.)_

ShadauxCat committed:
removed some unneeded checks on hashsets

ShadauxCat committed:
standards + fix broken test

ShadauxCat committed:
- Added `Not()` runtime targets
- Improved test coverage

ShadauxCat committed:
Standards

ShadauxCat committed:
Fix test failures in 2020.3

ShadauxCat committed:
Fix test failures in 2020.3 but for real this time

NoelStephensUnity committed:
Merge branch 'develop' into feat/universal_rpc_attribute

NoelStephensUnity committed:
Merge branch 'develop' into feat/universal_rpc_attribute

zachstronaut commented:
VERY excited for this, thank you @ShadauxCat

zachstronaut commented:
I wanted to test this, so opened this repository's testproject in Unity 2022.3.10f1 on Windows.

I tried adding a generic [Rpc] to GrabbableBall.cs

```
    [Rpc(SendTo.Everyone)]
    private void TestRpc()
    {
        Debug.Log("TestRpc");
    }
```

The first thing I noticed is that `[Rpc]` without arguments is not available.

When I ran the SampleScene as Host with no clients and tried to use my new RPC, I got the following error:

```
MethodAccessException: Method `Unity.Netcode.NetworkBehaviour.__beginSendRpc(uint,Unity.Netcode.RpcParams,Unity.Netcode.RpcAttribute/RpcAttributeParams,Unity.Netcode.SendTo,Unity.Netcode.RpcDelivery)' is inaccessible from method `GrabbableBall.TestRpc()'
GrabbableBall.TestRpc () (at Assets/Scripts/GrabbableBall.cs:113)
GrabbableBall.Update () (at Assets/Scripts/GrabbableBall.cs:64)
```

Would it be better to open a bug ticket for this?

zachstronaut commented:
I believe RuntimeAccessModifiersILPP is missing the following lines in the if condition in ProcessNetworkBehaviour()

```
methodDefinition.Name == nameof(NetworkBehaviour.__beginSendRpc) ||
methodDefinition.Name == nameof(NetworkBehaviour.__endSendRpc) ||
```

ShadauxCat commented:
@zachstronaut You are spot on. I'll get a PR with that fix open tomorrow. Thanks for catching that.

I'm also going to be integrating this into BossRoom tomorrow, which should help catch any other issues like this.
